### PR TITLE
Analysis

### PR DESCRIPTION
**Pull request by wizardfan**
introduce analysis data

### CODE REVIEW

File: scripts/validate_analysis_record.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    'updateDate',
+    'fileSizes',
+    'organism',
+    'urls',
+    'datasetInPortal'
+]
+logger = utils.create_logging_instance("validate_analysis")
+
+
+class validate_analysis_record(validate_record.validate_record):
```
Comment (Alexey-ebi) [inline scripts/validate_analysis_record.py]: class name should use CamelCase convention

File: scripts/validate_experiment_record.py [hunk 2]
```diff
@@ -1,250 +1,55 @@
...
+    'accession',  # validation tool use fixed id field
+    # not in the ruleset
+    'standardMet',
+    'versionLastStandardMet'
+]
+
+logger = utils.create_logging_instance('validate_experiment')
+
+
+class validate_experiment_record(validate_record.validate_record):
```
Comment (Alexey-ebi) [inline scripts/validate_experiment_record.py]: class name should use CamelCase convention

File: scripts/validate_record.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
+import requests
```
Comment (Alexey-ebi) [inline scripts/validate_record.py]: Unused import statement

File: scripts/validate_record.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+import sys
+from typing import Dict, List
+import utils
+from misc import from_lower_camel_case
+
+
+logger = utils.create_logging_instance("validate_record")
+
+
+class validate_record():
```
Comment (Alexey-ebi) [inline scripts/validate_record.py]: class name should use CamelCase convention

File: scripts/validate_record.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    def __init__(self, record_type: str, records: Dict, rulesets: List, batch_size: int = 600):
```
Comment (Alexey-ebi) [inline scripts/validate_record.py]: Please add class description and documentation for all methods inside it

File: scripts/change_alias.py [hunk 6]
```diff
@@ -1,13 +1,12 @@
...
 from elasticsearch import Elasticsearch
 import click
-import logging
 from utils import create_logging_instance, remove_underscore_from_end_prefix
 
 # logger = logging.getLogger(__name__)
 # logging.basicConfig(format='%(asctime)s\t%(levelname)s:\t%(name)s line %(lineno)s\t%(message)s', level=logging.INFO)
 # logging.getLogger('elasticsearch').setLevel(logging.WARNING)
-logger = create_logging_instance('change_alias', level=logging.INFO, to_file=False)
+logger = create_logging_instance('change_alias', to_file=False)
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/change_alias.py]: It is useful to get some feedback e.g. telling you the current indice used in alias. Would it be possible to redirect output in cron jobs?

File: scripts/import_analysis.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    :return:
+    """
+    hosts = es_hosts.split(";")
+    logger.info("Command line parameters")
+    logger.info("Hosts: "+str(hosts))
+    es_index_prefix = remove_underscore_from_end_prefix(es_index_prefix)
+    if es_index_prefix:
+        logger.info("Index_prefix:"+es_index_prefix)
+
+    es = Elasticsearch(hosts)
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: It is for printing parameters only, feels overkill to extract into functions

File: scripts/import_analysis.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                    file_es['checksum'] = analysis_es['checksums'][i]
+                    file_es['url'] = analysis_es['urls'][i]
+                    files_es.append(file_es)
+                analysis_es['files'] = files_es
+                analysis_es.pop('fileNames')
+                analysis_es.pop('fileTypes')
+                analysis_es.pop('fileSizes')
+                analysis_es.pop('checksumMethods')
+                analysis_es.pop('checksums')
+                analysis_es.pop('urls')
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: two factors made me to take this route
1. duplicating object requires extra resource
2. reaching the end of the loop, very certain that the analysis_es will not be used again

File: scripts/import_analysis_legacy.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        existing_datasets = get_datasets(hosts[0], es_index_prefix, only_faang=False)
+    except KeyError:
+        logger.error("No existing datasets retrieved from Elastic Search")
+        logger.error("Possible causes:")
+        logger.error("1 missing/wrong value for parameter es_index_prefix")
+        logger.error("2 ES server has connection issue, index does not exist etc.")
+        exit()
+
+    for study_accession in eva_datasets:
+        url = f"http://www.ebi.ac.uk/eva/webservices/rest/v1/studies/{study_accession}/summary"
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis_legacy.py]: eva api only used here, no need to extract

File: scripts/utils.py [hunk 10]
```diff
@@ -58,9 +60,90 @@ def insert_into_es(es, es_index_prefix, doc_type, doc_id, body):
...
+    es_doc['updateDate'] = record['last_updated']
+    es_doc.setdefault('organism', dict())
+    es_doc['organism']['text'] = record['scientific_name']
+    es_doc['organism']['ontologyTerms'] = f"http://purl.obolibrary.org/obo/NCBITaxon_{record['tax_id']}"
+
+    es_doc['datasetAccession'] = record['study_accession']
+    if record['study_accession'] in existing_datasets:
+        es_doc['datasetInPortal'] = True
+    else:
+        es_doc['datasetInPortal'] = False
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/utils.py]: only used once

File: scripts/import_analysis.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    hosts = es_hosts.split(";")
+    logger.info("Command line parameters")
+    logger.info("Hosts: "+str(hosts))
+    es_index_prefix = remove_underscore_from_end_prefix(es_index_prefix)
+    if es_index_prefix:
+        logger.info("Index_prefix:"+es_index_prefix)
+
+    es = Elasticsearch(hosts)
+
+    url = "https://www.ebi.ac.uk/ena/portal/api/search/?result=analysis&format=JSON&limit=0&fields=all&dataPortal=faang"
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: done in 2db90ac

File: scripts/import_analysis.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+from utils import remove_underscore_from_end_prefix, create_logging_instance, insert_into_es, get_datasets, \
+    convert_analysis
+from misc import get_filename_from_url
+import requests
+import json
+import validate_analysis_record
+from typing import List, Dict
+
+RULESETS = ["FAANG Analyses", "FAANG Legacy Analyses"]
+FILE_SERVER_TYPES = ['ftp', 'galaxy', 'aspera']
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: removed totally

File: scripts/utils.py [hunk 13]
```diff
@@ -58,9 +60,90 @@ def insert_into_es(es, es_index_prefix, doc_type, doc_id, body):
...
+        if only_faang:
+            if hit['_source']['standardMet'] == constants.STANDARD_FAANG:
+                datasets.add(hit['_id'])
+        else:
+            datasets.add(hit['_id'])
+    return datasets
+
+
+def convert_analysis(record, existing_datasets):
+    file_server_types = ['ftp', 'galaxy', 'aspera']
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/utils.py]: import_analysis has removed the code, feel no need to extract to constants

File: scripts/import_analysis_legacy.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        # expect always to have data from EVA as the list is retrieved live
+        # get EVA summary
+        eva_summary = requests.get(url).json()['response'][0]['result'][0]
+
+        url = f"https://www.ebi.ac.uk/ena/portal/api/search/?result=analysis&format=JSON&limit=0&" \
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis_legacy.py]: done in 2db90ac

File: scripts/import_analysis_legacy.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                es_doc['description'] = record['description']
+                if eva_summary['experimentType'] != '-':
+                    es_doc['experimentType'] = eva_summary['experimentType'].split(', ')
+                # es_doc['program'] = eva_summary['program']
+                if eva_summary['platform'] != '-':
+                    es_doc['platform'] = eva_summary['platform'].split(', ')
+                # imputation has not been exported in the ENA warehouse
+                # use PRJEB22988 (non farm animal) as example being both imputation and phasing project
+                # es_doc['imputation'] = record['imputation']
+            if es_doc:
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis_legacy.py]: fixed in 2db90ac

File: scripts/validate_analysis_record.py (hunk 1 above)
Comment (wizardfan) [reply scripts/validate_analysis_record.py]: fixed in 2db90ac

File: scripts/validate_experiment_record.py (hunk 2 above)
Comment (wizardfan) [reply scripts/validate_experiment_record.py]: fixed in 2db90ac

File: scripts/validate_record.py (hunk 3 above)
Comment (wizardfan) [reply scripts/validate_record.py]: done in 2db90ac

File: scripts/validate_record.py (hunk 4 above)
Comment (wizardfan) [reply scripts/validate_record.py]: fixed in 2db90ac

File: scripts/validate_record.py (hunk 5 above)
Comment (wizardfan) [reply scripts/validate_record.py]: done in b536f32

File: scripts/import_analysis.py [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+            es_doc['description'] = record['analysis_description']
+            es_doc['assayType'] = record['assay_type']
+            protocol = record['analysis_protocol']
+            es_doc.setdefault('analysisProtocol', dict())
+            es_doc['analysisProtocol']['url'] = protocol
+            es_doc['analysisProtocol']['filename'] = get_filename_from_url(protocol, record['analysis_accession'])
+            es_doc['referenceGenome'] = record['reference_genome']
+            # es_doc['analysisLink'] = record['analysis_alias']
+            # es_doc['analysisCodeRepository'] = record['analysis_alias']
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: fields in the ruleset, but no available real data, therefore warehouse could not generate the field, should be available in future

File: scripts/import_analysis.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+        if es_doc:
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: done in 2db90ac

File: scripts/import_analysis_legacy.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+    :return:
+    """
+    hosts = es_hosts.split(";")
+    logger.info("Command line parameters")
+    logger.info("Hosts: "+str(hosts))
+    es_index_prefix = remove_underscore_from_end_prefix(es_index_prefix)
+    if es_index_prefix:
+        logger.info("Index_prefix:"+es_index_prefix)
+
+    es = Elasticsearch(hosts)
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis_legacy.py]: same point as in import_analysis.py, only discuss there and close this comment

File: scripts/import_analysis.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: same apply in the import_analysis_legacy.py

File: scripts/import_analysis.py [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+                analysis_es.pop('fileNames')
+                analysis_es.pop('fileTypes')
+                analysis_es.pop('fileSizes')
+                analysis_es.pop('checksumMethods')
+                analysis_es.pop('checksums')
+                analysis_es.pop('urls')
+                body = json.dumps(analysis_es)
+                insert_into_es(es, es_index_prefix, 'analysis', analysis_accession, body)
+                # index into ES so break the loop
+                break
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: done in 073916d

File: scripts/import_analysis_legacy.py [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+                analysis_es.pop('fileNames')
+                analysis_es.pop('fileTypes')
+                analysis_es.pop('fileSizes')
+                analysis_es.pop('checksumMethods')
+                analysis_es.pop('checksums')
+                analysis_es.pop('urls')
+                body = json.dumps(analysis_es)
+                insert_into_es(es, es_index_prefix, 'analysis', analysis_accession, body)
+                # index into ES so break the loop
+                break
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis_legacy.py]: done in 073916d

File: scripts/import_analysis.py [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+            if not es_doc:
+                continue
+            es_doc['project'] = record['project_name']
+            es_doc.setdefault('experimentAccessions', list())
+            es_doc.setdefault('runAccessions', list())
+            for elmt in record['experiment_accession'].split(','):
+                if elmt:
+                    es_doc['experimentAccessions'].append(elmt)
+            for elmt in record['run_accession'].split(','):
+                if elmt:
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: done in 073916d

File: scripts/import_analysis.py [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+            es_doc = analyses[analysis_accession]
+        else:
+            es_doc = convert_analysis(record, existing_datasets)
+            if not es_doc:
+                continue
+            es_doc['project'] = record['project_name']
+            es_doc.setdefault('experimentAccessions', list())
+            es_doc.setdefault('runAccessions', list())
+            for elmt in record['experiment_accession'].split(','):
+                if elmt:
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis.py]: done in 073916d

File: scripts/import_analysis_legacy.py [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+
+
+def get_eva_dataset_list():
+    species_str = ",".join(EVA_SPECIES)
+    logger.info(f"Species to retrieve from EVA: {species_str}")
+    url = f"http://www.ebi.ac.uk/eva/webservices/rest/v1/meta/studies/all?species={species_str}"
```
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/import_analysis_legacy.py]: EVA API, only used once and could not see any chance to be used anywhere else.

File: scripts/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (wizardfan) [reply scripts/utils.py]: done in d8efb1f

**Pull request merged**